### PR TITLE
#1109 allow multiple environments to be given

### PR DESCRIPTION
**Pull request by ajoecker**
see https://github.com/getgauge/gauge/issues/1109

the change allows to give multiple environments to Gauge in the following way

`gauge run --env "foo, bar, blub"`

It follows the previous solution of first come, first serve, which means that if an env variable is set by a preceding env, it is not overwritten by a following one.

### CODE REVIEW

**Comment by sriv** [comment]
@ajoecker - I was thinking if this could be implemented in a better fashion. 

Perhaps have composable environments, so that

`foo_bar` would contain `foo` `union` `bar` + extra if needed. We still need to explore which takes precedence, but you get the idea.

Some benefits (IMO) : 

- the meta environments are also part of the project (i.e. checked in with tests) and can be replicated to other machines.
- allow extra properties to be defined (or even override property explicitly)

What do you think?

**Comment by ajoecker** [comment]
I'm not sure to get the idea of the composable environments - do you have a small example ?

**Comment by sriv** [comment]
Sure, here's what I think.

Say the project's `env` looks like this:

```

**Comment by sriv** [comment]
Sure, here's one.

Say the project's `env` looks like this:

```
env
├── bar
│   ├── default.properties
│   └── java.properties
├── default
│   ├── default.properties
│   └── java.properties
├── foo
│   ├── default.properties
│   └── java.properties
└── foo_bar
    └── inherited.properties
```

And the `foo_bar/inherited.properties` looks like this:

```
# include foo
# include bar

i_am_a_new_property = hello
```

Then `gauge run specs --env=foo_bar` should use the properties specified in `foo` and `bar`, as well as the new `i_am_a_new_property`

**Comment by ajoecker** [comment]
First, naiv thought.... if I look at the original issue, Thunderforge complains, that they must create additional folders/files to have a combination of environments available.

With your solution, I do not see a gain from that. If I want to combine `bar` and `foo`, I would again need to have additional work to do so. 
If you again look at Thunderforge's example, where he combines a lot of several environments, it would mean for each combination to have a union folder with different includes etc ?

Maybe I do not see the idea behind that right now, but to be able to combine on the fly, without much additional work, this does not seem so appealing.

I dont say that the idea in general is bad, this might be a interesting thought, but it does not help for this problem in my eyes (@Thunderforge might be able to add more to this ?!)

**Comment by sriv** [comment]
Let's continue the discussion on #1109

**Comment by sriv** [comment]
@gaugebot recheck the pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck the pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck this pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck this pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck this pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck this pull request.

**Comment by BugDiver** [comment]
@gaugebot recheck this pull request.

**Comment by ajoecker** [comment]
sorry - I missed that request... will take a look

File: env/env.go [hunk 1]
```diff
@@ -51,24 +51,29 @@ var currentEnv = "default"
...
+		currentEnv = strings.TrimSpace(currentEnv)
 
-	if currentEnv != "default" {
-		err := loadEnvDir("default")
+		err := loadEnvDir(currentEnv)
 		if err != nil {
 			return fmt.Errorf("Failed to load env. %s", err.Error())
 		}
+
+		if currentEnv != "default" {
```
[the remark this answers is not in the document]
Comment (ajoecker) [reply env/env.go]: loading default env is now outside the loop

**Comment by sriv** [comment]
`TestLoadMultipleEnvEnsureFirstOneDecides` is failing on travis - https://travis-ci.org/getgauge/gauge/jobs/414366676#L655

**Comment by haroon-sheikh** [comment]
@ajoecker @sriv the only and important use case I can see is to override a variable in env1 from env2. But from the changes, I can only see that it combines multiple environments by taking the first value. What decides the priority of the environments passed?

**Comment by ajoecker** [comment]
> @ajoecker @sriv the only and important use case I can see is to override a variable in env1 from env2. But from the changes, I can only see that it combines multiple environments by taking the first value. What decides the priority of the environments passed?

see discussion here https://github.com/getgauge/gauge/issues/1109#issuecomment-402888141

**Pull request merged**